Update `LightningModule` docs (#10637)
PyTorchLightning/pytorch-lightning: docs/source/common/lightning_module.rst

        new_x = torch.Tensor(2, 3)
        new_x = new_x.type_as(x)

> 5.  Lightning, by default, handles the distributed sampler for you when running under a distributed strategy.

Reviewer: I liked the "for you" statement more

Author: just trying to be a little more explicit about when actually lightning will handle distributed sampler.

> 5.  Lightning, by default, handles the distributed sampler for you when running under a distributed environment.

Author: ```suggestion
5.  Lightning, by default, handles the distributed sampler for you when running under a distributed strategy.
```

.. code-block:: python

    train_loader = DataLoader(MNIST(os.getcwd(), download=True, transform=transforms.ToTensor()))
    trainer = pl.Trainer(max_epochs=1)
    model = LitModel()

>     trainer.fit(model, train_dataloaders=train_loader)

Author: its `train_dataloaders`

     - the full validation loop
   * - test_step
     - the full test loop

>    * - predict_step

Reviewer 2: Thanks for adding the missing predict references.

Train epoch-level operations
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> If you need to do something with all the outputs of each :meth:`~pytorch_lightning.core.lightning.LightningModule.training_step`.

Author: updated!

Training with DataParallel
~~~~~~~~~~~~~~~~~~~~~~~~~~

> When training using an ``accelerator`` that splits data from each batch across GPUs, sometimes you might

Reviewer 2: ```suggestion
When training using an ``strategy`` that splits data from each batch across GPUs, sometimes you might
```

need to aggregate them on the main GPU for processing (dp, or ddp2).

> In this case, implement the :meth:`~pytorch_lightning.core.lightning.LightningModule.training_step_end`.

Author: updated!

         for out in training_step_outputs:
             ...

> The complete pseudocode that lightning does under the hood is:

Author: sure.. will take care for others too!

            torch.set_grad_enabled(True)
            model.train()

You can also run just the validation loop on your val_dataloaders by overriding :meth:`~pytorch_lightning.core.lightning.LightningModule.validation_step`
and calling :meth:`~pytorch_lightning.trainer.trainer.Trainer.validate`.

.. code-block:: python

    model = Model()
    trainer = Trainer()
    trainer.validate(model)

Validation epoch-level metrics
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
If you need to do something with all the outputs of each :meth:`~pytorch_lightning.core.lightning.LightningModule.validation_step`,

> override :meth:`~pytorch_lightning.core.lightning.LightningModule.validation_epoch_end`

Author: done!

Validating with DataParallel
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> When training using an ``strategy`` that splits data from each batch across GPUs, sometimes you might

Reviewer 3: ```suggestion
When training using a ``strategy`` that splits data from each batch across GPUs, sometimes you might
```

        task = ClassificationTask(model)

        trainer = Trainer(gpus=2)

>         trainer.fit(task, train_dataloaders=train_dataloader, val_dataloaders=val_dataloader)

Author: are you sure?
https://github.com/PyTorchLightning/pytorch-lightning/blob/ce0a977742872736f150f6d37ecaa301a318668f/pytorch_lightning/trainer/trainer.py#L693-L700

.. automethod:: pytorch_lightning.core.lightning.LightningModule.save_hyperparameters
    :noindex:

> toggle_optimizer

Reviewer: It's used in some code examples, i'd say yes.

global_rank
~~~~~~~~~~~

> The global_rank of this LightningModule. Lightning saves logs, weights etc. only from global_rank = 0. You
> usually do not need to use this property

Reviewer 4: ```suggestion
The ``global_rank`` is the index of the current process across all nodes and devices.
Lightning will perform some operations such as logging, weight checkpointing only when ``global_rank is 0``. You
usually do not need to use this property, but it is useful to know how to access it if needed.
```

> Global rank refers to the index of that GPU across ALL GPUs. For example, if using 10 machines, each with 4 GPUs,

Reviewer 4: ```suggestion
For example, if using 10 machines, each with 4 GPUs,
```

hparams
~~~~~~~

> The arguments saved by calling :meth:`~pytorch_lightning.core.mixins.HyperparametersMixin.save_hyperparameters`

Reviewer 2: ```suggestion
The arguments saved by calling ``save_hyperparameters``
```
I see you did multiple changes to this format, I prefer the existing one. Imo, user doesn't need to know the reference

Author: aren't references better? if they want to have a look at it.. then it's quick, no search required.

local_rank
~~~~~~~~~~~

> The local_rank of this LightningModule. Lightning saves logs, weights etc. only from global_rank = 0. You

Reviewer 4: ```suggestion
The ``local_rank`` of this LightningModule. Lightning saves logs, weights etc. only from global_rank = 0. You
```

Reviewer 4: I will reformulate this. We are talking about `local_rank` and global_rank is used as above. This is confusing.

Author: done!

truncated_bptt_steps
^^^^^^^^^^^^^^^^^^^^

> Truncated back prop breaks perform backprop every k steps of

Reviewer 4: ```suggestion
Truncated backpropagation through time performs backpropagation every k steps of
```
